keps/sig-api-machinery/0014-garbage-collection.md

```diff
@@ -0,0 +1,555 @@
+* Because etcd TTL is used, it does not respect other lifecycle guarantees,
+  such as finalizers.
+
+#### "Root" Resources Without an Owner
+
+Resources that are not terminated Pods, Events, or controllers history
+resources, can only be garbage collected with ownerReference-based garbage
+collection. 
+
+However, OwnerReference-based garbage collection can only handle the case when a
+resource has at least an owner. However, not all resources have owners (i.e.
+non-empty `.metadata.ownerReferences`). You can think of the resources that
+don't have owners as root nodes in a DAG (directed acyclic graph). Those
+resources without owners may never be deleted after they're no longer needed. 
+
+Most of the Kubernetes resources that users create directly won't have owners. 
+
+#### Immutable Pattern: Short-lived, Frequently-created Resources
+
+Short-lived resources, such as Jobs (that are not managed by CronJobs), Secrets
+and ConfigMaps, are often treated as _immutable_ resources -- they are
+frequently created and not updated by users. We also have discussed making such
+resources immutable ([#10179][]).
+
+However, it's difficult for the users to clean up those resources automatically,
+because those resources don't have owners. Without a way to clean up those
+frequently created, short-lived resources, those resources can accumulate and
+overload a Kubernetes cluster very easily. Even if we can avoid the overload
+issue by implementing a cluster-wide (global) resource quota, users won't be
+able to create new resources without cleaning up old ones first. 
+
+[#10179]: https://github.com/kubernetes/kubernetes/issues/10179#issuecomment-384019058
+
+#### Cannot Centralize Cleanup of Multiple Resources with OwnerReferences
+
+Moreover, people building declarative application description with several
+Kubernetes resources cannot easily centralize clean up using owner references,
+meaning, to create a bunch of objects and deleting one of them to make the other
+objects to go away with it. 
+
+One of the reasons is that owner resource [UID][] is required when specifying
+owner references, but UID is unknown until after a resource is created. Even if
+Kubernetes removes this limitation and allow named based owner references,
+there’s still a resource creation ordering problem: any child resource created
+before its parent resource is prone to be cleaned up by the garbage collector
+unexpectedly.
+
+What's more, a child resource may depend on many other parent resources. It's
+likely for the list of owners of a child resource to be excessively long, and
+hit Kubernetes resource size limit. The resource then cannot be updated anymore
+and becomes useless. 
+
+[UID]: https://kubernetes.io/docs/concepts/overview/working-with-objects/names/#uids
+
+## Proposal
+
+The design has three main parts: 
+
+1. **Object-lifecycle-respecting TTL mechanism.** Provide observability before
+   deletion, and address the resource creation ordering problems. See [detailed
```

Comment:
> address the resource creation ordering problems

it wasn't clear to me how this addressed that... was it because the TTL gives you a window within which the owning objects can be created? what would we recommend someone wanting to create a set of interdependent objects without worrying about creation order set this to? a few seconds? a few minutes?

Reply:
The main problem I want to solve is the observability before deletion. We can solve the creation ordering problem with something more reliable, such as having GC controller to remember if it has ever observed the owner or not. 

If we don't want to encourage using TTLs to solve problems like creation order, we can make TTLs to be "GC after use" only. We can even combine TTL and GCIfUnused fields. For example, have a "durability" field, which can be `durable` (default, don't GC after use) or `ephemeral` (GC after not used for 2d).